### PR TITLE
Experimental mixin optimisation

### PR DESCRIPTION
**Pull request by ForbesLindesay**
I need to take a lot more time reviewing and simplifying this code
before I'm happy with it, but it does something pretty powerful.  It
makes mixins and mixin blocks be functions that return strings.  It
then hoists all mixins that don't refer to any locals/global variables.
 Finally, for all mixins where there is a 1-1 mapping between mixin
name and code, it updates the reference to directly refer to the
function, rather than going via the jade_mixins object.

For example:

mixin.attrs.jade

``` jade
mixin centered(title)
  div.centered(id=attributes.id)
    - if (title)
      h1(class=attributes.class)= title
    block
    - if (attributes.href)
      .footer
        a(href=attributes.href) Back

mixin main(title)
  div.stretch
    +centered(title).highlight&attributes(attributes)
      block

mixin bottom
  div.bottom&attributes(attributes)
    block

body
  +centered#First Hello World
  +centered('Section 1')#Second
    p Some important content.
  +centered('Section 2')#Third.foo(href='menu.html', class='bar')
    p Even more important content.
  +main('Section 3')(href='#')
    p Last content.
  +bottom.foo(class='bar', name='end', id='Last', data-attr='baz')
    p Some final words.
  +bottom(class=['class1', 'class2'])

mixin foo
  div.thing(attr1='foo', attr2='bar')&attributes(attributes)

- var val = '<biz>'
- var classes = ['foo', 'bar']
+foo(attr3='baz' data-foo=val data-bar!=val class=classes).thunk

//- Regression test for #1424
mixin work_filmstrip_item(work)
  div&attributes(attributes)= work
+work_filmstrip_item('work')("data-profile"='profile', "data-creator-name"='name')

+foo(
  attr3="qux"
  class="baz"
)
```

before:

``` js
function jade_attr(e, t, n, r) {
    if (("class" === e || "style" === e) && !t) return "";
    if ("boolean" == typeof t || null == t) return t ? " " + (r ? e : e + '="' + e + '"') : "";
    if (t && "function" == typeof t.toISOString) t = t.toISOString(); else if ("string" != typeof t && (t = JSON.stringify(t), 
    !n && -1 !== t.indexOf('"'))) return " " + e + "='" + t.replace(/'/g, "&apos;") + "'";
    return n ? " " + e + '="' + jade_escape(t) + '"' : " " + e + '="' + t + '"';
}

function jade_attrs(e, a) {
    for (var s = [], t = Object.keys(e), r = 0; r < t.length; ++r) {
        var j = t[r], l = e[j];
        "class" == j && (l = jade_classes(l)), "style" === j && (l = jade_style(l)), s.push(jade_attr(j, l, !1, a));
    }
    return s.join("");
}

function jade_classes(e, a) {
    return Array.isArray(e) && Array.isArray(a) ? jade_classes(e.map(jade_classes).map(function(e, r) {
        return a[r] ? jade_escape(e) : e;
    })) : (Array.isArray(e) ? e.map(jade_classes) : e && "object" == typeof e ? Object.keys(e).filter(function(a) {
        return e[a];
    }) : [ e ]).filter(Boolean).join(" ");
}

function jade_escape(e) {
    var a = ("" + e).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
    return a === "" + e ? e : a;
}

function jade_merge(e, r) {
    if (1 === arguments.length) {
        for (var a = e[0], t = 1; t < e.length; t++) a = jade_merge(a, e[t]);
        return a;
    }
    for (var n in r) if ("class" === n) {
        var s = e[n] || [];
        e[n] = (Array.isArray(s) ? s : [ s ]).concat(r[n] || []);
    } else "style" === n ? (e[n] = jade_style(e[n]), r[n] = jade_style(r[n]), e[n] = e[n] + (e[n] && r[n] && ";") + r[n]) : e[n] = r[n];
    return e;
}

function jade_style(e) {
    return e && "object" == typeof e ? Object.keys(e).map(function(t) {
        return t + ":" + e[t];
    }).join(";") : ("" + (e || "")).replace(/\;$/, "");
}

function template(locals) {
    var buf = [], jade_mixins = {}, jade_interp;
    jade_mixins["centered"] = function(title) {
        var block = this && this.block, attributes = this && this.attributes || {};
        buf.push("<div" + jade_attr("id", attributes.id, true, false) + ' class="centered">');
        if (title) {
            buf.push("<h1" + jade_attr("class", jade_classes([ attributes.class ], [ true ]), false, false) + ">" + jade_escape(null == (jade_interp = title) ? "" : jade_interp) + "</h1>");
        }
        block && block();
        if (attributes.href) {
            buf.push('<div class="footer"><a' + jade_attr("href", attributes.href, true, false) + ">Back</a></div>");
        }
        buf.push("</div>");
    };
    jade_mixins["main"] = function(title) {
        var block = this && this.block, attributes = this && this.attributes || {};
        buf.push('<div class="stretch">');
        jade_mixins["centered"].call({
            block: function() {
                block && block();
            },
            attributes: jade_merge([ {
                "class": "highlight"
            }, attributes ])
        }, title);
        buf.push("</div>");
    };
    jade_mixins["bottom"] = function() {
        var block = this && this.block, attributes = this && this.attributes || {};
        buf.push("<div" + jade_attrs(jade_merge([ {
            "class": "bottom"
        }, attributes ]), false) + ">");
        block && block();
        buf.push("</div>");
    };
    buf.push("<body>");
    jade_mixins["centered"].call({
        block: function() {
            buf.push("Hello World");
        },
        attributes: {
            id: "First"
        }
    });
    jade_mixins["centered"].call({
        block: function() {
            buf.push("<p>Some important content.</p>");
        },
        attributes: {
            id: "Second"
        }
    }, "Section 1");
    jade_mixins["centered"].call({
        block: function() {
            buf.push("<p>Even more important content.</p>");
        },
        attributes: {
            id: "Third",
            href: "menu.html",
            "class": "foo bar"
        }
    }, "Section 2");
    jade_mixins["main"].call({
        block: function() {
            buf.push("<p>Last content.</p>");
        },
        attributes: {
            href: "#"
        }
    }, "Section 3");
    jade_mixins["bottom"].call({
        block: function() {
            buf.push("<p>Some final words.</p>");
        },
        attributes: {
            name: "end",
            id: "Last",
            "data-attr": "baz",
            "class": "foo bar"
        }
    });
    jade_mixins["bottom"].call({
        attributes: {
            "class": "class1 class2"
        }
    });
    buf.push("</body>");
    jade_mixins["foo"] = function() {
        var block = this && this.block, attributes = this && this.attributes || {};
        buf.push("<div" + jade_attrs(jade_merge([ {
            attr1: "foo",
            attr2: "bar",
            "class": "thing"
        }, attributes ]), false) + "></div>");
    };
    var val = "<biz>";
    var classes = [ "foo", "bar" ];
    jade_mixins["foo"].call({
        attributes: {
            attr3: "baz",
            "data-foo": typeof (jade_interp = val) == "string" ? jade_escape(jade_interp) : jade_interp,
            "data-bar": val,
            "class": jade_classes([ classes, "thunk" ], [ true, false ])
        }
    });
    jade_mixins["work_filmstrip_item"] = function(work) {
        var block = this && this.block, attributes = this && this.attributes || {};
        buf.push("<div" + jade_attrs(attributes, false) + ">" + jade_escape(null == (jade_interp = work) ? "" : jade_interp) + "</div>");
    };
    jade_mixins["work_filmstrip_item"].call({
        attributes: {
            "data-profile": "profile",
            "data-creator-name": "name"
        }
    }, "work");
    jade_mixins["foo"].call({
        attributes: {
            attr3: "qux",
            "class": "baz"
        }
    });
    return buf.join("");
}
```

after:

``` js
function jade_attr(e, t, n, r) {
    if (("class" === e || "style" === e) && !t) return "";
    if ("boolean" == typeof t || null == t) return t ? " " + (r ? e : e + '="' + e + '"') : "";
    if (t && "function" == typeof t.toISOString) t = t.toISOString(); else if ("string" != typeof t && (t = JSON.stringify(t), 
    !n && -1 !== t.indexOf('"'))) return " " + e + "='" + t.replace(/'/g, "&apos;") + "'";
    return n ? " " + e + '="' + jade_escape(t) + '"' : " " + e + '="' + t + '"';
}

function jade_attrs(e, a) {
    for (var s = [], t = Object.keys(e), r = 0; r < t.length; ++r) {
        var j = t[r], l = e[j];
        "class" == j && (l = jade_classes(l)), "style" === j && (l = jade_style(l)), s.push(jade_attr(j, l, !1, a));
    }
    return s.join("");
}

function jade_classes(e, a) {
    return Array.isArray(e) && Array.isArray(a) ? jade_classes(e.map(jade_classes).map(function(e, r) {
        return a[r] ? jade_escape(e) : e;
    })) : (Array.isArray(e) ? e.map(jade_classes) : e && "object" == typeof e ? Object.keys(e).filter(function(a) {
        return e[a];
    }) : [ e ]).filter(Boolean).join(" ");
}

function jade_escape(e) {
    var a = ("" + e).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
    return a === "" + e ? e : a;
}

function jade_merge(e, r) {
    if (1 === arguments.length) {
        for (var a = e[0], t = 1; t < e.length; t++) a = jade_merge(a, e[t]);
        return a;
    }
    for (var n in r) if ("class" === n) {
        var s = e[n] || [];
        e[n] = (Array.isArray(s) ? s : [ s ]).concat(r[n] || []);
    } else "style" === n ? (e[n] = jade_style(e[n]), r[n] = jade_style(r[n]), e[n] = e[n] + (e[n] && r[n] && ";") + r[n]) : e[n] = r[n];
    return e;
}

function jade_style(e) {
    return e && "object" == typeof e ? Object.keys(e).map(function(t) {
        return t + ":" + e[t];
    }).join(";") : ("" + (e || "")).replace(/\;$/, "");
}

var jade_mixins = {}, jade_interp;

function jade_mixin_56b9486ac45949e031ec70f28720d245dbd16e53_centered(jade_mixin_data, title) {
    var buf = [];
    var block = jade_mixin_data && jade_mixin_data.block;
    var attributes = jade_mixin_data && jade_mixin_data.attributes || {};
    buf.push("<div" + jade_attr("id", attributes.id, true, false) + ' class="centered">');
    if (title) {
        buf.push("<h1" + jade_attr("class", jade_classes([ attributes.class ], [ true ]), false, false) + ">" + jade_escape(null == (jade_interp = title) ? "" : jade_interp) + "</h1>");
    }
    buf.push(block ? block() : "");
    if (attributes.href) {
        buf.push('<div class="footer"><a' + jade_attr("href", attributes.href, true, false) + ">Back</a></div>");
    }
    buf.push("</div>");
    return buf.join("");
}

function jade_mixin_7709fe0c84f69612760199424903216fce507b04_bottom(jade_mixin_data) {
    var buf = [];
    var block = jade_mixin_data && jade_mixin_data.block;
    var attributes = jade_mixin_data && jade_mixin_data.attributes || {};
    buf.push("<div" + jade_attrs(jade_merge([ {
        "class": "bottom"
    }, attributes ]), false) + ">" + (block ? block() : "") + "</div>");
    return buf.join("");
}

function jade_mixin_84c9badee8226f73edc8f828c883b502bcb88f56_foo(jade_mixin_data) {
    var buf = [];
    var attributes = jade_mixin_data && jade_mixin_data.attributes || {};
    buf.push("<div" + jade_attrs(jade_merge([ {
        attr1: "foo",
        attr2: "bar",
        "class": "thing"
    }, attributes ]), false) + "></div>");
    return buf.join("");
}

function jade_mixin_d5e0c32019d17ad1ae63b795fa5e1578da81cbb9_work_filmstrip_item(jade_mixin_data, work) {
    var buf = [];
    var attributes = jade_mixin_data && jade_mixin_data.attributes || {};
    buf.push("<div" + jade_attrs(attributes, false) + ">" + jade_escape(null == (jade_interp = work) ? "" : jade_interp) + "</div>");
    return buf.join("");
}

function template(locals) {
    var buf = [];
    function jade_mixin_017269efd73f32e4838d066a0585949475f28a32_main(jade_mixin_data, title) {
        var buf = [];
        var block = jade_mixin_data && jade_mixin_data.block;
        var attributes = jade_mixin_data && jade_mixin_data.attributes || {};
        buf.push('<div class="stretch">' + jade_mixin_56b9486ac45949e031ec70f28720d245dbd16e53_centered({
            block: function() {
                var buf = [];
                buf.push(block ? block() : "");
                return buf.join("");
            },
            attributes: jade_merge([ {
                "class": "highlight"
            }, attributes ])
        }, title) + "</div>");
        return buf.join("");
    }
    buf.push("<body>" + jade_mixin_56b9486ac45949e031ec70f28720d245dbd16e53_centered({
        block: function() {
            var buf = [];
            buf.push("Hello World");
            return buf.join("");
        },
        attributes: {
            id: "First"
        }
    }) + jade_mixin_56b9486ac45949e031ec70f28720d245dbd16e53_centered({
        block: function() {
            var buf = [];
            buf.push("<p>Some important content.</p>");
            return buf.join("");
        },
        attributes: {
            id: "Second"
        }
    }, "Section 1") + jade_mixin_56b9486ac45949e031ec70f28720d245dbd16e53_centered({
        block: function() {
            var buf = [];
            buf.push("<p>Even more important content.</p>");
            return buf.join("");
        },
        attributes: {
            id: "Third",
            href: "menu.html",
            "class": "foo bar"
        }
    }, "Section 2") + jade_mixin_017269efd73f32e4838d066a0585949475f28a32_main({
        block: function() {
            var buf = [];
            buf.push("<p>Last content.</p>");
            return buf.join("");
        },
        attributes: {
            href: "#"
        }
    }, "Section 3") + jade_mixin_7709fe0c84f69612760199424903216fce507b04_bottom({
        block: function() {
            var buf = [];
            buf.push("<p>Some final words.</p>");
            return buf.join("");
        },
        attributes: {
            name: "end",
            id: "Last",
            "data-attr": "baz",
            "class": "foo bar"
        }
    }) + jade_mixin_7709fe0c84f69612760199424903216fce507b04_bottom({
        attributes: {
            "class": "class1 class2"
        }
    }) + "</body>");
    var val = "<biz>";
    var classes = [ "foo", "bar" ];
    buf.push(jade_mixin_84c9badee8226f73edc8f828c883b502bcb88f56_foo({
        attributes: {
            attr3: "baz",
            "data-foo": typeof (jade_interp = val) == "string" ? jade_escape(jade_interp) : jade_interp,
            "data-bar": val,
            "class": jade_classes([ classes, "thunk" ], [ true, false ])
        }
    }));
    buf.push(jade_mixin_d5e0c32019d17ad1ae63b795fa5e1578da81cbb9_work_filmstrip_item({
        attributes: {
            "data-profile": "profile",
            "data-creator-name": "name"
        }
    }, "work") + jade_mixin_84c9badee8226f73edc8f828c883b502bcb88f56_foo({
        attributes: {
            attr3: "qux",
            "class": "baz"
        }
    }));
    return buf.join("");
}
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -393,86 +457,144 @@ Compiler.prototype = {
...
+    }
+    if (pp) this.buf.push("jade_indent.pop();")
+  },
+  visitMixinDefinition: function (mixin) {
+    var originalCalledMixins = this.calledMixins;
+    this.calledMixins = [];
+    var args = mixin.args ? mixin.args.split(',') : [];
+    args.unshift('jade_mixin_data');
+    var rest;
+    if (/^\.\.\../.test(args[args.length - 1].trim())) {
```
Comment (TimothyGu) [inline index.js:539]: Extra dot?

``` diff
-    if (/^\.\.\./.test(args[args.length - 1].trim())) {
+    if (/^\.\.\../.test(args[args.length - 1].trim())) {
```

Comment (ForbesLindesay) [inline index.js:539]: There should be at least some character for the variable name, in addition to the `...` to mark it as a rest param.  We should probably be more restrictive and actually error early if any of these are not valid identifiers.  I don't want to get hung up on that detail at this point though.  The higher level optimisation is much more important.

**Comment by jeromew** [comment]
I am not sure about the impact of such an optimization but here is my use case for mixins :
I use the jade_mixins object to inject external functions into jade. In jade templates, I use dynamic mixin calls `+#{name}(..)` to call these functions. This way I can compile my jade files independently from the compilation of the mixins ; I can even hot-swap a mixin implementation without recompiling jade.

I fear that optimizations will prevent such componentification of mixins but maybe I misunderstand the pr

File: index.js (hunk 1 above)
Comment (TimothyGu) [inline index.js:539]: Okay, just noticed that detail and was wondering if it was on purpose.

**Comment by jeromew** [comment]
I see that you use a `jade_mixin_data` in the prototype of the mixins instead of the old `.call({})` trick + that the `buf` is now autonomous (instead of coming from the closure).

I think this is a lot better for modularization if mixins can use zero variables from the parent scope but I wonder if the good API for this is

```
var s = mixin(jade_mixin_data/*block, attributes*/,..)
```

or

```
mixin(jade_mixin_data/*buf, block, attributes*/,..)
```

Have you also thought about making a mixin call async ? It seems that the default jade scheme is synchronous for a reason ;

`then-jade` can always override this and I know there is a discussion on `buf.push` versus string concatenation but for sure I will need access to the `buf` in order to allow streaming from within the mixins.

Another option might also be in `then-jade` for mixins to output a stream or use something as `stream-splicer` to build the stream graph upon mixin or block invocations.

I don't know if this is disturbing to have `then-jade` talks on those API discussions ; I know that you did not want  `then-jade` to impact thoughts on jade 1.x ; I am not so sure about jade 2.x maybe you can clarify this for me. thanks

**Comment by ForbesLindesay** [comment]
I think this is best left to the babel rewrite.

**Pull request closed without merging**